sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,550 @@
+   * sent (`sending` == true) or shall be the next to send. May be NULL iff
+   * this stream has no completed buffers to send.
+   */
+  usb_logging_buf_t *send_buf;
+  /**
+   * Time at which the most recently-populated buffer should be sent to the
+   * host; prevent log data being retained indefinitely in the event of no
+   * further log output.
+   */
+  ibex_timeout_t flush_time;
+  /**
+   * The current logging buffer; may be partially-filled or full (see
+   * `bytes_used` field), or NULL iff this stream has no pending logging data.
+   */
+  usb_logging_buf_t *wr_buf;
+} usb_logging_stream_t;
+
+// USB Logging context.
+struct usb_logging_ctx {
+  /**
+   * Access to usb_testutils layer.
+   */
+  usb_testutils_ctx_t *usbutils;
+  /**
+   * Was the logging responsible for initializing the testutils layer?
+   */
+  bool inited;
+  /**
+   * Number of logging streams.
+   */
+  uint8_t nstreams;
+  /**
+   * Linked-list of free buffers.
+   */
+  usb_logging_buf_t *free;
+  /**
+   * Per-stream state information.
+   */
+  usb_logging_stream_t streams[kUSBLogMaxStreams];
+  /**
+   * Pool of logging buffers; shared among all active streams.
+   */
+  usb_logging_buf_t buf[kUSBLogNumBuffers];
+};
+
+/**
+ * USB device context types.
+ */
+static usb_testutils_ctx_t usbdev;
+static usb_testutils_controlep_ctx_t usbdev_control;
+
+// Note: for now there is a single logging context.
+static usb_logging_ctx_t the_ctx;
+
+/**
+ * Pinmux handle
+ */
+static dif_pinmux_t pinmux;
+
+/**
+ * Internal development logging; be careful about using this if stdout has been
+ * redirected!
+ */
+static const bool kDevLogging = false;
+
+/**
+ * Character map, used to ensure that non-printable characters are dropped;
+ * isprint() available so this assumes ASCII and drops the control characters
+ * that could could issues with inappropriate terminal settings at the host.
+ */
+static const uint32_t kCtrlChars =
+    ((uint32_t)1u << '\t') | ((uint32_t)1u << '\n') | ((uint32_t)1u << '\r');
+static const uint32_t kMap[0x100u / 0x20u] = {
+    // suppress control characters and ASCII 127 (Delete)
+    kCtrlChars, ~0u, ~0u, 0x7fffffffu,
+    // top-bit set characters are harmless; propagate them unchanged.
+    ~0u, ~0u, ~0u, ~0u};
+
+// Our stdout function
+static size_t usb_log(void *data, const char *buf, size_t len);
+
+/**
+ * Configuration values for USB.
+ */
+static uint8_t config_descriptors[kCfgDscrLenMax];
+
+// Since isprint() is unavailable, use a simple bitmap to test whether it is
+// safe to transmit the given ASCII character unmodified.
+static bool isprintable(uint8_t ch) {
+  return ((kMap[ch >> 5] >> (ch & 0x1fu)) & 1u) != 0u;
+}
+
+// Place all buffers into the free list
+static void free_list_init(usb_logging_ctx_t *ctx) {
+  usb_logging_buf_t *prev = NULL;  // End of list.
+  unsigned idx = kUSBLogNumBuffers;
+  // Order them such that the first buffer is initially at the head.
```

Comment:
This could just be a for loop?
```c
for (size_t i = 1; i < kUSBLogNumBuffers; i++) {
  ctx->buf[idx - 1].next = &ctx->buf[idx];
}
ctx->buf[kUSBLogNumBuffers - 1].next = NULL;
```

Reply:
Sure. The comment was just in anticipation of someone questioning why the loop runs down rather than up. I presume we're okay with while() loops, generally? It is more efficient simply to remember the pointer from the previous iteration.

Reply:
The compiler should be quite good with this strength reduction. Generally it's good to iterating up to down due to better caching behaviour (although non-existent on ibex).

Reply:
That's common subexpression elimination (CSE) rather than strength reduction, but why write more code in the first place, and then rely upon both the human reader and the compiler spotting that relationship between the current element and its neighbor?
Cacheing behavior/performance is unaffected unless there is prefetching, and I'd be surprised if there's a prefetcher that assumes ascending addresses but cannot handle descending just as well.